### PR TITLE
Global leaderboard `add_posts_to_main_feed`

### PR DESCRIPTION
**Pull request by hlbmtc**
**TODO:**

- [ ] Migrate all private projects to have `add_to_main_feed=False`
- [x] Ensure global leaderboard only includes questions with default **PUBLIC** project

related #1657

### CODE REVIEW

File: scoring/models.py [hunk 1]
```diff
@@ -157,6 +157,75 @@ def get_questions(self) -> list[Question]:
...
+                    checked_intervals.append((start, end))
+                    questions = questions.filter(
+                        Q(open_time__lt=start)
+                        | Q(scheduled_close_time__gt=end + close_grace_period)
+                        | Q(actual_resolve_time__gt=end + resolve_grace_period)
+                    )
+            questions = questions.exclude(
+                related_posts__post__curation_status__in=invalid_statuses
+            ).distinct("pk")
+            return list(questions)
```
Comment (lsabor) [inline scoring/models.py:227]: This was a major rework of how questions are fetched. It reduced the call to a fraction of a second from ~2-3 seconds locally.
I verified that the results are the same as the previous method of iterating through each question and checking it's global leaderboard dates.
To my understanding, this was the big reason why global leaderboard contributions were a weak point.

This doesn't speed up comment insight / question writing contributions, though. That's for another time.

File: scoring/models.py [hunk 2]
```diff
@@ -157,6 +157,75 @@ def get_questions(self) -> list[Question]:
...
         ]
 
+        if self.project and self.project.type == Project.ProjectTypes.SITE_MAIN:
+            # global leaderboard
+            if self.start_time is None or self.end_time is None:
+                raise ValueError("Global leaderboards must have start and end times")
+            questions = Question.objects.filter(
+                Q(related_posts__post__projects__add_posts_to_main_feed=True)
+                | Q(related_posts__post__default_project__add_posts_to_main_feed=True)
+            )
```
Comment (hlbmtc) [inline scoring/models.py:167]: I made a small performance testing and this takes `5s` on my local machine.
Though, this one takes `0.8s` and _does not require_ `.distinct()`:
```python
Question.objects.filter(
    related_posts__post__in=Post.objects.filter_for_main_feed()
)
```

File: scoring/models.py [hunk 3]
```diff
@@ -157,6 +157,75 @@ def get_questions(self) -> list[Question]:
...
+
+            if self.score_type == self.ScoreTypes.COMMENT_INSIGHT:
+                # post must be published
+                return list(
+                    questions.filter(
+                        related_posts__post__published_at__lt=self.end_time
+                    ).distinct("pk")
```
Comment (hlbmtc) [inline scoring/models.py:174]: Remove distinct after

File: scoring/models.py [hunk 4]
```diff
@@ -157,6 +157,75 @@ def get_questions(self) -> list[Question]:
...
+                    questions.filter(
+                        Q(scheduled_close_time__gte=self.start_time)
+                        & (
+                            Q(actual_close_time__isnull=True)
+                            | Q(actual_close_time__gte=self.start_time)
+                        ),
+                        related_posts__post__published_at__lt=self.end_time,
+                    )
+                    .exclude(related_posts__post__curation_status__in=invalid_statuses)
+                    .distinct("pk")
```
Comment (hlbmtc) [inline scoring/models.py:189]: Same

File: scoring/models.py (hunk 2 above)
Comment (lsabor) [reply scoring/models.py:167]: Wow, 5 seconds is not good. Were you testing the endpoint response time, or just the `get_questions` method?
But either way, down to 0.8s is a crazy improvement.
The thing is, I was getting the whole `get_questions` method to run in `0.3s` locally when it was `2-3s` before the change.
We can talk through the details when we sync today.

**Comment by hlbmtc** [comment]
related #1621

**Pull request merged**